### ISSUE TITLE
perf: comprehensive CPU performance audit - beat TorchSharp/TensorFlow across all operations

### ISSUE THREAD

**Issue by ooples** [ooples/AiDotNet.Tensors#25]
# CPU Performance Comprehensive Audit Report

## Executive Summary

A deep analysis of all CPU tensor operations reveals **systemic performance bottlenecks** across the entire CpuEngine. While AiDotNet.Tensors beats TorchSharp on Sum (1.3x faster) and Mean (1.5x faster), it loses significantly on most other operations:

| Operation | AiDotNet | TorchSharp | Gap | Priority |
|-----------|----------|------------|-----|----------|
| **ReLU** (1M) | 102 us | 15.5 us | **6.5x slower** | P0 |
| **Sigmoid** (1M) | 1,131 us | 121 us | **9.4x slower** | P0 |
| **MatMul** (256x256) | 184 us | 84.5 us | **2.2x slower** | P0 |
| **MatMul** (512x512) | 817 us | 440 us | **1.9x slower** | P0 |
| **Conv2D** (16ch 64x64) | 318 us | 238 us | **1.3x slower** | P1 |
| **Add** (1M) | 62 us | 35.3 us | **1.8x slower** | P0 |
| **Multiply** (1M) | 62 us | 32.8 us | **1.9x slower** | P0 |
| Sum (1M) | 23.5 us | 31.6 us | 1.3x FASTER | P3 |
| Mean (1M) | 25.8 us | 38.9 us | 1.5x FASTER | P3 |

**Root causes fall into 5 categories:**
1. **ToArray() epidemic** - 324 `.ToArray()` calls across the codebase causing 3 allocations + 3 copies per operation
2. **Missing SIMD** - 496 scalar `numOps` call sites vs only 29 `TensorPrimitivesHelper` calls
3. **Missing parallelization** - Most operations are single-threaded even for 1M+ elements
4. **BLAS underutilization** - Only 2 of 16+ matmul operations attempt BLAS dispatch
5. **Infrastructure overhead** - Lock contention on engine access, false sharing, inconsistent thresholds

---

### 1. TensorPrimitivesHelper: Eliminate 26 ToArray() Sites

**File:** `src/AiDotNet.Tensors/Helpers/TensorPrimitivesHelper.cs`
**Impact:** Every vector operation does 3 allocations + 3 copies instead of zero-copy span operations

**Current (broken):**
```csharp
public static Vector<T> Add(Vector<T> x, Vector<T> y)
{
    var xArray = x.ToArray();   // ALLOC + COPY
    var yArray = y.ToArray();   // ALLOC + COPY
    var result = new T[xArray.Length];  // ALLOC
    NumOps.Add(xArray, yArray, result);
    return new Vector<T>(result);   // COPIES again
}
```

**Fix:** Use `x.AsSpan()` / `x.AsMemory()` for zero-copy:
```csharp
public static Vector<T> Add(Vector<T> x, Vector<T> y)
{
    var result = new Vector<T>(x.Length);
    NumOps.Add(x.AsSpan(), y.AsSpan(), result.AsWritableSpan());
    return result;
}
```

**26 methods to fix** - all follow same pattern (Add, Subtract, Multiply, Divide, Negate, Abs, Sqrt, etc.)

---

### 2. ReLU: Add Parallelization + In-Place SIMD

**Files:** `CpuEngine.cs:4317`, `FloatOperations.cs:1632`, `SimdKernels.cs:294`
**Impact:** 6.5x slower than TorchSharp (102 us vs 15.5 us for 1M elements)

**Root cause:** Single-threaded AVX loop, no parallelization for large tensors.

**Current call chain:**
```
CpuEngine.ReLUInPlace -> numOps.ReLU -> SimdKernels.ReLU (single-threaded AVX)
```

**Fix:** Add parallel chunking in FloatOperations.ReLU:
```csharp
public void ReLU(ReadOnlySpan<float> x, Span<float> destination)
{
    if (x.Length >= ParallelThreshold)
    {
        ParallelForChunks(x.Length, MinChunkSize, (start, count) =>
        {
            SimdKernels.ReLU(x.Slice(start, count), destination.Slice(start, count));
        });
    }
    else
    {
        SimdKernels.ReLU(x, destination);
    }
}
```

Same fix needed for: LeakyReLU, GELU (already vectorized in SimdKernels but not parallelized).

---

### 3. Sigmoid: Eliminate 3 Unnecessary Memory Copies

**File:** `FloatOperations.cs:1146`
**Impact:** 9.4x slower than TorchSharp (1,131 us vs 121 us for 1M elements)

**Current (3 unnecessary copies):**
```csharp
var xArray = pool.Rent(length);      // Rent from pool
var destArray = pool.Rent(length);   // Rent from pool
x.CopyTo(xArray);                   // COPY 1: span -> rented array
ParallelForChunks(length, MinChunkSize, (start, count) => {
    TensorPrimitives.Sigmoid(xArray.AsSpan(start, count), destArray.AsSpan(start, count));
});
destArray.AsSpan(0, length).CopyTo(destination);  // COPY 2: rented -> destination
pool.Return(xArray);
pool.Return(destArray);
```

**Fix:** Operate directly on spans for in-place:
```csharp
public void Sigmoid(ReadOnlySpan<float> x, Span<float> destination)
{
    if (x.Length >= ParallelThreshold)
    {
        ParallelForChunks(x.Length, MinChunkSize, (start, count) =>
        {
            TensorPrimitives.Sigmoid(x.Slice(start, count), destination.Slice(start, count));
        });
    }
    else
    {
        TensorPrimitives.Sigmoid(x, destination);
    }
}
```

**Same ArrayPool+copy pattern exists in:** GELU, Mish, Swish, ELU (5 total sites in FloatOperations).

---

### 4. MatMul Fallback: Eliminate params int[] Heap Allocations

**File:** `CpuEngine.cs:5051`
**Impact:** 2.2x slower than TorchSharp (184 us vs 84.5 us for 256x256)

**Root cause:** When BLAS fails to load, fallback uses `a[i, k]` / `b[k, j]` indexer which allocates `params int[]` on every access. For 256x256 matmul: ~50 million heap allocations.

**Current (broken fallback):**
```csharp
Parallel.For(0, m, i =>
{
    for (int j = 0; j < p; j++)
    {
        T sum = numOps.Zero;
        for (int k = 0; k < n; k++)
            sum = numOps.Add(sum, numOps.Multiply(a[i, k], b[k, j])); // params int[] alloc!
        result[i, j] = sum;
    }
});
```

**Fix:** Use `GetFlat()` with pre-computed linear indices:
```csharp
Parallel.For(0, m, i =>
{
    int aRowStart = i * n;
    for (int j = 0; j < p; j++)
    {
        T sum = numOps.Zero;
        for (int k = 0; k < n; k++)
            sum = numOps.Add(sum, numOps.Multiply(
                a.GetFlat(aRowStart + k), b.GetFlat(k * p + j)));
        result.SetFlat(i * p + j, sum);
    }
});
```

**Also:** Verify BLAS is actually loading - add diagnostic output to benchmarks. All BLAS failures are silently swallowed in `BlasProvider.cs`.

---

### 5. BLAS Underutilization: Only 2 of 16+ MatMul Ops Use BLAS

**Files:** `CpuEngine.cs`, `MatrixMultiplyHelper.cs`
**Impact:** Most matrix operations bypass BLAS entirely

**Operations that SHOULD use BLAS but do not:**
- `BatchMatMul` 2D (line 1790) - No BLAS, no parallelism, no SIMD at all
- `FusedGemmBiasActivation` - Claims "SIMD dot product" but uses scalar loop
- `DotProduct` (line 363) - Uses scalar loop instead of existing `SimdKernels.DotProduct`
- `TensorMatMulBatched` (line 5118) - Calls `.ToArray()` on entire tensors
- All attention Q@K^T operations - Scalar triple loops
- All normalization matmul operations

**Fix:** Route through `MatrixMultiplyHelper.TryGemm` for all float/double matmul operations. For batch operations, parallelize the batch dimension and use BLAS for each batch element.

---

### 6. AiDotNetEngine.Current Lock Contention

**File:** `src/AiDotNet.Tensors/Engines/AiDotNetEngine.cs`
**Impact:** Every engine access takes a lock - scalability bottleneck under parallelism

**Fix:** Replace `lock` with `Volatile.Read`/`Volatile.Write` or `Interlocked.CompareExchange` pattern.

---

### 7. Element-wise Operations: Scalar GetFlat Loops

**File:** `CpuEngine.cs`
**Impact:** Add/Multiply 1.8-1.9x slower than TorchSharp

**Operations using scalar `GetFlat()` loops instead of TensorPrimitives:**
- `TensorDivide` - scalar loop
- `TensorMultiplyScalar` - scalar loop
- `TensorSubtract` - scalar loop
- `TensorAbs` - scalar loop when `TensorPrimitives.Abs` exists
- `TensorLog` - scalar loop when `TensorPrimitives.Log` exists
- `TensorExp` - scalar loop when `TensorPrimitives.Exp` exists
- `TensorSqrt` - scalar loop when `TensorPrimitives.Sqrt` exists

**Fix:** Use `TensorPrimitives.*` on contiguous spans with parallel chunking for large tensors.

---

### 8. Mish/Swish/ELU: Zero SIMD Vectorization

**File:** `SimdKernels.cs:509-566`
**Impact:** These activations use scalar loops even though vectorized approximations exist

```csharp
// Current - pure scalar
public static void Mish(ReadOnlySpan<float> input, Span<float> output)
{
    for (int i = 0; i < input.Length; i++)
        output[i] = input[i] * MathF.Tanh(MathF.Log(1f + MathF.Exp(input[i])));
}
```

**Fix:** Use TensorPrimitives for Mish/Swish (available in .NET 9+) or implement AVX+FMA vectorized versions similar to the existing GELU implementation.

---

### 9. All Backward Passes: ToArray + Scalar ToDouble/FromDouble

**File:** `CpuEngine.cs:17298-17391`
**Impact:** 10-20x slower than forward passes, affects all training

```csharp
// Current pattern in ALL backward passes:
var inputArray = input.ToArray();    // Full copy
var gradArray = outputGrad.ToArray(); // Full copy
for (int i = 0; i < totalElements; i++)
{
    double x = numOps.ToDouble(inputArray[i]);    // Virtual dispatch
    double g = numOps.ToDouble(gradArray[i]);     // Virtual dispatch
    double result = g * SomeDerivative(x);
    gradInput.SetFlat(i, numOps.FromDouble(result)); // Virtual dispatch
}
```

**Fix:** For float tensors, use TensorPrimitives directly. For generic, at minimum use span access instead of ToArray.

---

### 10. Derivative Functions: Per-Element Parallel.For

**File:** `CpuEngine.cs:14365-14421`
**Impact:** Creates one Task per element instead of chunking

```csharp
// Current - one task per element!
Parallel.For(0, totalElements, i =>
{
    double x = numOps.ToDouble(input.GetFlat(i));
    result.SetFlat(i, numOps.FromDouble(SigmoidDerivative(x)));
});
```

**Fix:** Use `ParallelForChunks` with proper chunk sizes, or better yet, use TensorPrimitives for float specialization.

---

### 11. Conv2D Dispatch Order: SIMD Before BLAS

**File:** `CpuEngine.cs` (Conv2D dispatch)
**Impact:** SIMD direct convolution runs before im2col+BLAS GEMM, which is faster for most sizes

**Current order:** SIMD direct conv -> im2col+BLAS GEMM
**Correct order:** im2col+BLAS GEMM -> SIMD direct conv (for small kernels only)

---

### 12. TensorMaxValue/MinValue/SumOfSquares: Scalar When TensorPrimitives Exists

**File:** `CpuEngine.cs`

- `TensorMaxValue`: Scalar comparisons -> use `TensorPrimitives.Max`
- `TensorMinValue`: Scalar comparisons -> use `TensorPrimitives.Min`
- `TensorSumOfSquares`: `ToArray()` + scalar loop -> use `TensorPrimitives.SumOfSquares`

---

### 13. All Normalization Ops: Zero SIMD

**File:** `CpuEngine.cs`
**Operations:** BatchNorm, LayerNorm, GroupNorm, InstanceNorm, RMSNorm

All use scalar `numOps` calls with `.ToArray()` copies. For float tensors, these should use TensorPrimitives for mean/variance computation and element-wise normalization.

---

### 14. All Pooling Ops: No SIMD, No Parallelization

**File:** `CpuEngine.cs`
**Operations:** MaxPool2D, AvgPool2D, AdaptiveAvgPool2D

All use nested loops with tensor indexer (params int[] allocation). Should use flat span access with SIMD comparisons for max, and TensorPrimitives.Sum for average pooling.

---

### 15. Missing In-Place Activation Variants

9 activations exist only as copy operations, not in-place:
- Softplus, HardSwish, HardSigmoid, SELU, CELU, Softsign, LogSigmoid, Softmin, Sparsemax

Adding in-place variants would eliminate allocation for these operations.

---

### 16. False Sharing in Parallel Sum

**File:** `FloatOperations.cs`
**Impact:** Partial sums stored in shared array cause cache line bouncing

**Fix:** Pad partial sum array entries to cache line boundaries (64 bytes apart).

---

### 17. MinChunkSize Inconsistency

- `CpuParallelSettings.cs`: MinChunkSize = 4,096
- `FloatOperations.cs`: MinChunkSize = 8,192

Should be consistent. 8,192 is more appropriate for SIMD operations (aligns with typical L1 cache lines).

---

### 18. Conv2D Double: Zero Optimization Strategies

**File:** `CpuEngine.cs`
Float Conv2D has 4 strategies (oneDNN, SIMD direct, im2col+BLAS, Winograd). Double has zero - falls through to pure scalar loops.

At minimum, double should get im2col+BLAS GEMM support.

---

### 19. Vector ReLU Bypasses SimdKernels

**File:** `CpuEngine.cs`
When operating on `Vector<T>` (not `Tensor<T>`), ReLU uses a scalar loop instead of routing through SimdKernels.ReLU.

---

### 20. Loss Functions: Completely Unoptimized

All loss functions use `.ToArray()` + `numOps.ToDouble()` per element. For float tensors, most could use TensorPrimitives directly (Sum, Multiply, Log, Abs, etc.).

---

### 21. Attention Q@K^T: Scalar Triple Loops
Should use BLAS GEMM for the matrix multiply portion.

### 22. Double Significantly Worse Than Float
Float has TensorPrimitives specialization; double should too (TensorPrimitives supports double in .NET 9+).

### 23. Further Optimize Where We Already Win
Sum and Mean already beat TorchSharp but could be even faster with:
- Eliminating false sharing in partial sums
- Using non-temporal stores for large reductions

---

## Complete Operation Inventory

**Total public methods in CpuEngine:** ~460
**Methods using scalar numOps:** 496 call sites
**Methods using TensorPrimitivesHelper:** 29 call sites
**Total .ToArray() calls across codebase:** 324

### By Category:
| Category | Methods | SIMD Coverage | Priority |
|----------|---------|---------------|----------|
| Activations (forward) | 37 | 4/37 (11%) | P0-P1 |
| Activations (backward) | 37 | 0/37 (0%) | P1 |
| MatMul variants | 16+ | 2/16 (13%) | P0 |
| Element-wise ops | 20+ | 3/20 (15%) | P0-P1 |
| Reductions | 8 | 2/8 (25%) | P1-P2 |
| Normalization | 5+ | 0/5 (0%) | P1 |
| Pooling | 3+ | 0/3 (0%) | P1 |
| Conv2D | 4 strategies | 2/4 (50%) | P1 |
| Loss functions | 15+ | 0/15 (0%) | P2 |
| Attention | 3+ | 0/3 (0%) | P2-P3 |

---

## Recommended Fix Order

1. **TensorPrimitivesHelper ToArray elimination** (26 sites) - Unblocks all vector operations
2. **FloatOperations Sigmoid/GELU/Mish copies** (5 sites) - Fix 9.4x gap
3. **FloatOperations ReLU parallelization** - Fix 6.5x gap
4. **MatMul fallback indexer** - Fix 2.2x gap
5. **BLAS diagnostics in benchmarks** - Verify BLAS loads
6. **Element-wise TensorPrimitives** (7+ ops) - Fix 1.8x gap
7. **AiDotNetEngine.Current lock** - Fix scalability
8. **Backward pass optimization** - Fix training speed
9. **Normalization SIMD** - Fix BatchNorm/LayerNorm
10. **Pooling SIMD** - Fix MaxPool/AvgPool

---

## Verification Plan

After each fix:
1. Run `dotnet test` to verify correctness
2. Run `--vs-torchsharp-cpu` benchmark to measure improvement
3. A/B test: compare one change at a time
4. Target: beat TorchSharp on ALL operations
